### ISSUE TITLE
[data grid] Improve keyboard navigation based on the ARIA spec

### ISSUE THREAD

**Issue by arminmeh** [mui/mui-x#14204]
### Summary

Data grid already supports [keyboard interaction](https://mui.com/x/react-data-grid/accessibility/#keyboard-navigation), which is aligned with the [grid keyboard interaction ARIA specs](https://www.w3.org/WAI/ARIA/apg/patterns/grid/), but it is not aligned with the [ARIA specs](https://www.w3.org/WAI/ARIA/apg/patterns/treegrid/#keyboard_interaction) for the treegrid

Among other things, data tree grid is missing expand/collapse control via arrow keys

From the spec
> Right Arrow:
If focus is on a collapsed row, expands the row.
If focus is on an expanded row or is on a row that does not have child rows, moves focus to the first cell in the row.
If focus is on the right-most cell in a row, focus does not move.
If focus is on any other cell, moves focus one cell to the right.
Left Arrow:
If focus is on an expanded row, collapses the row.
If focus is on a collapsed row or on a row that does not have child rows, focus does not move.
If focus is on the first cell in a row and row focus is supported, moves focus to the row.
If focus is on the first cell in a row and row focus is not supported, focus does not move.
If focus is on any other cell, moves focus one cell to the left.

Similar ticket for tree view: #12654

**Search keywords**: accessibility, keyboard, aria
